### PR TITLE
TOOLS-1163 Support dumping and restoring collections with slashes in their names

### PR DESCRIPTION
**Pull request by patrickfreed**
[TOOLS-1163](https://jira.mongodb.com/browse/TOOLS-1163)

This PR fixes a bug in mongodump/mongorestore where collections with slashes in their names could not be dumped/restored. It does so by URL encoding the collection name before writing to / reading from disk.

Because we're using URL encoding and not just special casing slashes, the default filenames will change for other cases, too. 

e.g. If "mydb" contains collections "my/coll" and "my coll":
`mongodump -d mydb` will create a directory with `my%20coll.bson` and `my+coll.bson` in it.

This also means that the filenames that mongorestore will be importing from will be unescaped. So if you have an old dump of a collection with a % in its name, restoring from it after this patch will fail or restore to an unexpected collection name.

e.g.
`mongorestore mydb/my%coll.bson` would fail
`mongorestore mydb/my%21coll.bson` would restore to `mydb.my!coll` instead of `mydb.my%21coll`.

In mongoexport, the default is to write to stdout, so the user needs to specify an outfile if they want it written to disk. Therefore, collections with slashes are already supported, and the filesystem-compatible name issues are offloaded to the user. 

In mongoimport, it will put the documents in a collection based on the filename of the provided data, so you cannot import to a collection with slashes in its name without specifying the name yourself with `-c`. Other than that, collections with slashes in their names are fully supported. 

So, in both mongoimport/mongoexport there are no changes. We could add url encoding parsing to mongoimport, but that would require the user to manually url encode their outfile names in mongoexport, which I don't think is a good idea.

Also, I've put some changes into the `/common` directory of the tools repository. Once this is approved, I can merge those changes into the `common` repository instead and update this patch to point to there instead.

### CODE REVIEW

File: mongodump/prepare.go [hunk 1]
```diff
@@ -193,10 +194,13 @@ func (dump *MongoDump) outputPath(dbName, colName string) string {
 	} else {
 		root = dump.OutputOptions.Out
 	}
+
+	escapedCollName := util.EscapeCollectionName(colName)
+
 	if dbName == "" {
-		return filepath.Join(root, colName)
+		return filepath.Join(root, escapedCollName)
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply mongodump/prepare.go]: fixed

File: mongorestore/filepath.go [hunk 2]
```diff
@@ -206,34 +206,40 @@ func (f *stdinFile) Close() error {
...
-		baseName := strings.TrimSuffix(baseFileName, ".bson")
-		return baseName, BSONFileType
+		baseName = strings.TrimSuffix(baseFileName, ".bson")
+		fileType = BSONFileType
 	}
-	return "", UnknownFileType
+
+	unescaped, err := util.UnescapeCollectionName(baseName)
+	if err != nil {
+		return "", UnknownFileType, fmt.Errorf("error parsing collection name from filename: %v", err)
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply mongorestore/filepath.go]: yep, fixed.

File: mongodump/prepare.go [hunk 3]
```diff
@@ -193,10 +194,8 @@ func (dump *MongoDump) outputPath(dbName, colName string) string {
 	} else {
 		root = dump.OutputOptions.Out
 	}
-	if dbName == "" {
-		return filepath.Join(root, colName)
-	}
-	return filepath.Join(root, dbName, colName)
+
+	return filepath.Join(root, dbName, util.EscapeCollectionName(colName))
```
Comment (xdg) [inline mongodump/prepare.go]: Is there significance to eliminating the case when `dbName` is empty?

Comment (xdg) [reply mongodump/prepare.go]: I also wonder if we should be percent-encoding the database name.

File: common/util/file.go [hunk 4]
```diff
@@ -40,6 +41,14 @@ func ToUniversalPath(path string) string {
 	return filepath.FromSlash(path)
 }
 
+func EscapeCollectionName(collName string) string {
+	return url.QueryEscape(collName)
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply common/util/file.go]: Done

File: common/util/file.go [hunk 5]
```diff
@@ -40,6 +41,14 @@ func ToUniversalPath(path string) string {
...
+}
+
+func UnescapeCollectionName(escapedCollName string) (string, error) {
+	return url.QueryUnescape(escapedCollName)
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply common/util/file.go]: Done

File: mongodump/prepare.go (hunk 3 above)
Comment (patrickfreed) [reply mongodump/prepare.go]: >Is there significance to eliminating the case when dbName is empty?

It's just a bit of cleanup. As per @kmahar 's [comment](https://github.com/mongodb/mongo-tools/pull/126#discussion_r261770219), `filepath.Join` ignores empty strings.

>I also wonder if we should be percent-encoding the database name.

Database names are restricted from having `/\. "$` in them, so any valid db name shouldn't have to be escaped (I think).

**Pull request merged**